### PR TITLE
BASE-1268-configured-quantiles

### PR DESCRIPTION
**Pull request by veronikaKochugova**

### CODE REVIEW

File: src/main/java/com/emc/mongoose/load/step/client/LoadStepClientBase.java [hunk 1]
```diff
@@ -54,9 +47,15 @@
...
+import static com.emc.mongoose.Constants.KEY_CLASS_NAME;
+import static com.emc.mongoose.Constants.KEY_STEP_ID;
+import static com.emc.mongoose.config.ConfigUtil.flatten;
+import static org.apache.logging.log4j.CloseableThreadContext.Instance;
+import static org.apache.logging.log4j.CloseableThreadContext.put;
+
 public abstract class LoadStepClientBase
-extends LoadStepBase
-implements LoadStepClient {
+	extends LoadStepBase
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/load/step/client/LoadStepClientBase.java]: is it possible to configure the IDE for the project's code style?

File: src/main/resources/config/defaults.json [hunk 2]
```diff
@@ -1,561 +1,559 @@
...
+			"trace" : {
 				// Persist the information about each load operation if true
-				"persist": false
+				"persist" : false
 			},
 			// The concurrency threshold to enable intermediate statistics calculation,
 			// 0 means no threshold
-			"threshold": 0
+			"threshold" : 0,
+			"quantiles" : [
```
Comment (akurilov) [inline src/main/resources/config/defaults.json]: but where is a median??

File: src/main/java/com/emc/mongoose/metrics/MetricsManagerImpl.java [hunk 3]
```diff
@@ -175,7 +177,7 @@ public void unregister(final MetricsContext metricsCtx) {
 						Loggers.METRICS_STD_OUT.info(new StepResultsMetricsLogMessage(distributedMetricsCtx));
 						final PrometheusMetricsExporter exporter = distributedMetrics.remove(distributedMetricsCtx);
 						if(exporter != null) {
-							CollectorRegistry.defaultRegistry.unregister(exporter);
+							CollectorRegistry.defaultRegistry.unregister((PrometheusMetricsExporterImpl) exporter);
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/metrics/MetricsManagerImpl.java]: it's possible just to cast to Collector

File: src/main/java/com/emc/mongoose/metrics/util/PrometheusMetricsExporterImpl.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	private final List<Double> quantileValues = new ArrayList<>();
+	private String help = "";
+
+	public PrometheusMetricsExporterImpl(final DistributedMetricsContext context) {
+		this.metricsContext = context;
+	}
+
+	@Override
+	public PrometheusMetricsExporterImpl quantile(final double value) {
+		if(value <= 1.0 && value >= 0.0) {
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/metrics/util/PrometheusMetricsExporterImpl.java]: quantile == 1 is not in the valid range

File: src/main/java/com/emc/mongoose/load/step/client/LoadStepClientBase.java (hunk 1 above)
Comment (veronikaKochugova) [reply src/main/java/com/emc/mongoose/load/step/client/LoadStepClientBase.java]: I tried many times, but it does not work out at all( and I can’t even manually remove this indent.

File: src/main/resources/config/defaults.json (hunk 2 above)
Comment (veronikaKochugova) [reply src/main/resources/config/defaults.json]: the median is always calculated and there is a separate method for it.

File: src/main/java/com/emc/mongoose/load/step/client/LoadStepClientBase.java [hunk 5]
```diff
@@ -339,20 +338,24 @@ protected final void initMetrics(
 		final boolean metricsAvgPersistFlag = metricsConfig.boolVal("average-persist");
 		final boolean metricsSumPersistFlag = metricsConfig.boolVal("summary-persist");
 		final boolean metricsSumPerfDbOutputFlag = metricsConfig.boolVal("summary-perfDbResultsFile");
+		final List<Double> quantileValues = metricsConfig.listVal("quantiles")
```
[the remark this answers is not in the document]
Comment (veronikaKochugova) [reply src/main/java/com/emc/mongoose/load/step/client/LoadStepClientBase.java]: `final List<String> storageNodeAddrs = storageNetNodeConfig.listVal("addrs");`
I do not see the difference

[the remark this answers is not in the document]
Comment (akurilov) [reply src/main/java/com/emc/mongoose/load/step/client/LoadStepClientBase.java]: I was wrong, that is fine

File: src/main/java/com/emc/mongoose/load/step/client/LoadStepClientBase.java (hunk 1 above)
Comment (akurilov) [reply src/main/java/com/emc/mongoose/load/step/client/LoadStepClientBase.java]: ok

File: src/main/resources/config/defaults.json (hunk 2 above)
Comment (veronikaKochugova) [reply src/main/resources/config/defaults.json]: Ok

File: src/test/java/com/emc/mongoose/integration/ExposedMetricsTest.java [hunk 6]
```diff
@@ -71,14 +71,17 @@ public void setUp()
...
+			new double[] {}
 		);
 		snapshotsSupplier = () -> Arrays.asList(metricsContext.lastSnapshot());
 		metricsContext.start();
 		//
 		distributedMetricsContext = new DistributedMetricsContextImpl(
 			STEP_ID, OP_TYPE, nodeCountSupplier, concurrencyLimit, concurrencyThreshold, ITEM_DATA_SIZE,
-			UPDATE_INTERVAL_SEC, true, true, true, true, snapshotsSupplier
+			UPDATE_INTERVAL_SEC, true, true, true, true, snapshotsSupplier,
+			new double[] {}
```
Comment (akurilov) [inline src/test/java/com/emc/mongoose/integration/ExposedMetricsTest.java]: Please add some custom quantiles into this test

**Pull request closed without merging**